### PR TITLE
Fix RTPS submessage counts when bundling alters outbound submessage order

### PR DESCRIPTION
**Pull request by simpsont-oci**

### CODE REVIEW

**Comment by mitza-oci** [comment]
Looks like there is one test failure on the Jenkins GHPRB build that is new to this branch

File: dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp [hunk 1]
```diff
@@ -2283,16 +2284,23 @@ RtpsUdpDataLink::bundle_mapped_meta_submessages(AddrDestMetaSubmessageMap& adr_m
           }
         }
         // Attempt to add the submessage meta_submessage to the bundle
-        bool result = false;
+        bool result = false, unique = false;
+        ACE_UNUSED_ARG(unique);
```
Comment (mitza-oci) [inline dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp:2288]: Why is this needed?

File: dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp [hunk 2]
```diff
@@ -2283,16 +2284,23 @@ RtpsUdpDataLink::bundle_mapped_meta_submessages(AddrDestMetaSubmessageMap& adr_m
...
         MetaSubmessage& res = **resp_it;
         switch (res.sm_._d()) {
           case HEARTBEAT: {
+            const EntityId_t id = res.sm_.heartbeat_sm().writerId;
+            unique = counts.heartbeat_counts_[id].insert(res.sm_.heartbeat_sm().count.value).second;
+            OPENDDS_ASSERT(unique == true);
```
Comment (mitza-oci) [inline dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp:2294]:
```suggestion
            OPENDDS_ASSERT(unique);
```

File: dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp [hunk 3]
```diff
@@ -2368,6 +2380,29 @@ RtpsUdpDataLink::send_bundled_submessages(MetaSubmessageVec& meta_submessages)
         std::memcpy(&idst.guidPrefix, dst.guidPrefix, sizeof(idst.guidPrefix));
         ser << idst;
       }
+      switch (res.sm_._d()) {
+        case HEARTBEAT: {
+          CountSet& set = counts.heartbeat_counts_[res.sm_.heartbeat_sm().writerId];
+          res.sm_.heartbeat_sm().count.value = *(set.begin());
```
Comment (mitza-oci) [inline dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp:2386]:
```suggestion
          res.sm_.heartbeat_sm().count.value = *set.begin();
```

File: dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp (hunk 1 above)
Comment (simpsont-oci) [reply dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp:2288]: I assumed that if we were setting unique below but never referencing it when the asserts were optimized out, some compilers would complain it was unused / unreferenced.

Comment (mitza-oci) [reply dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp:2288]: I missed the detail that this was assigned-to but not read-from.

File: dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp [hunk 4]
```diff
@@ -2368,6 +2380,29 @@ RtpsUdpDataLink::send_bundled_submessages(MetaSubmessageVec& meta_submessages)
         std::memcpy(&idst.guidPrefix, dst.guidPrefix, sizeof(idst.guidPrefix));
         ser << idst;
       }
+      switch (res.sm_._d()) {
+        case HEARTBEAT: {
+          CountSet& set = counts.heartbeat_counts_[res.sm_.heartbeat_sm().writerId];
```
Comment (jrw972) [inline dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp:2385]: Do you want to `OPENDDS_ASSERT(!set.empty());`?

**Review by jrw972 — APPROVED** [review]
